isobus/src/can_transport_protocol.cpp

```diff
@@ -11,860 +11,1004 @@
-		                                                      data.data(),
-		                                                      8,
-		                                                      source,
-		                                                      destination,
-		                                                      CANIdentifier::CANPriority::PriorityDefault6);
-	}
+		bool foundSession = false;
 
-	void TransportProtocolManager::close_session(TransportProtocolSession *session, bool successfull)
-	{
-		if (nullptr != session)
+		if (auto session = get_session(source, destination))
 		{
-			process_session_complete_callback(session, successfull);
-			auto sessionLocation = std::find(activeSessions.begin(), activeSessions.end(), session);
-			if (activeSessions.end() != sessionLocation)
+			if (session->get_parameter_group_number() == pgn)
 			{
-				activeSessions.erase(sessionLocation);
-				delete session;
-				CANStackLogger::CAN_stack_log(CANStackLogger::LoggingLevel::Debug, "[TP]: Session Closed");
+				foundSession = true;
+				CANStackLogger::error("[TP]: Received an abort (reason=%hu) for an rx session for PGN 0x%05X", static_cast<uint8_t>(reason), pgn);
+				close_session(*session, false);
+			}
+		}
+		if (auto session = get_session(destination, source))
+		{
+			if (session->get_parameter_group_number() == pgn)
+			{
+				foundSession = true;
+				CANStackLogger::error("[TP]: Received an abort (reason=%hu) for a tx session for PGN 0x%05X", static_cast<uint8_t>(reason), pgn);
+				close_session(*session, false);
 			}
 		}
-	}
 
-	void TransportProtocolManager::process_session_complete_callback(TransportProtocolSession *session, bool success)
-	{
-		if ((nullptr != session) &&
-		    (nullptr != session->sessionCompleteCallback) &&
-		    (nullptr != session->sessionMessage.get_source_control_function()) &&
-		    (ControlFunction::Type::Internal == session->sessionMessage.get_source_control_function()->get_type()))
+		if (!foundSession)
 		{
-			session->sessionCompleteCallback(session->sessionMessage.get_identifier().get_parameter_group_number(),
-			                                 session->get_message_data_length(),
-			                                 std::static_pointer_cast<InternalControlFunction>(session->sessionMessage.get_source_control_function()),
-			                                 session->sessionMessage.get_destination_control_function(),
-			                                 success,
-			                                 session->parent);
+			CANStackLogger::warn("[TP]: Received an abort (reason=%hu) with no matching session for PGN 0x%05X", static_cast<uint8_t>(reason), pgn);
 		}
 	}
 
-	bool TransportProtocolManager::send_broadcast_announce_message(TransportProtocolSession *session) const
+	void TransportProtocolManager::process_connection_management_message(const CANMessage &message)
 	{
-		bool retVal = false;
+		const auto pgn = message.get_uint24_at(5);
```

Comment:
Should maybe validate the message size first, otherwise could throw if the length is 1?

Reply:
Hmm, it had a check for this already one layer up. I moved it and made it an early-return to hopefully give a better overview next time